File: http/src/main/java/com/mackwu/http/rxjava/BaseObserver.java
package com.mackwu.http.rxjava;

import com.mackwu.http.bean.BaseResponse;

import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.net.UnknownServiceException;
import java.util.List;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import retrofit2.HttpException;

/**
 * ===================================================
 * Created by MackWu on 2020/6/29 19:05
 * <a href="mailto:wumengjiao828@163.com">Contact me</a>
 * <a href="https://github.com/mackwu828">Follow me</a>
 * ===================================================
 */
public abstract class BaseObserver<T> implements Observer<BaseResponse<T>>, IOnResult<T> {

    private Disposable disposable;

    @Override
    public void onSubscribe(final Disposable d) {
        this.disposable = d;
    }

    @Override
    public void onNext(final BaseResponse<T> response) {
        if (response.getError() != 0) {
            onFail("error: " + response.getError() + ", message: " + response.getMessage());
            return;
        }
        T data = response.getData();
        if (null == data) {
            onFail("data is null");
            return;
        }
        if (data instanceof List && ((List) data).isEmpty()) {
            onFail("data is empty");
            return;
        }
        onSuccess(data, response.getMessage());
    }

    @Override
    public void onError(final Throwable e) {
        String message = e.getMessage();

        if (e instanceof HttpException) {

        } else if (e instanceof SocketTimeoutException) {

        } else if (e instanceof RuntimeException) {

        } else if (e instanceof UnknownHostException) {

        } else if (e instanceof UnknownServiceException) {

        } else {

        }
        onFail(message);
        dispose();
    }

    @Override
    public void onComplete() {
        dispose();
    }

    private void dispose() {
        if (null != disposable && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }
}

File: http/src/main/java/com/mackwu/http/okhttp_retrofit_rxjava/BaseObserver.kt
//package com.mackwu.http.okhttp
//
//
//import com.zeasn.accu.services.bean.BaseResponse
//import io.reactivex.Observer
//import io.reactivex.disposables.Disposable
//import retrofit2.HttpException
//import java.net.SocketTimeoutException
//import java.net.UnknownHostException
//import java.net.UnknownServiceException
//
///**
// * ===================================================
// * Created by MackWu on 2018/5/24 11:46
// * <a href="mailto:wumengjiao828@163.com">Contact me</a>
// * <a href="https://github.com/mackwu828">Follow me</a>
// * ===================================================
// */
//abstract class BaseObserver<T> : Observer<BaseResponse<T>> {
//
//    private var disposable: Disposable? = null
//
//    abstract fun onSuccess(t: T?, successMsg: String)
//
//    abstract fun onFail(errorMsg: String)
//
//    override fun onSubscribe(d: Disposable) {
//        disposable = d
//    }
//
//    override fun onNext(response: BaseResponse<T>) {
//        response.run {
//            if (error == 0) {
//                onSuccess(datas, "")
//                return
//            }
//            onFail("")
//        }
//    }
//
//    override fun onError(t: Throwable) {
//        t.printStackTrace()
//        val errorMsg = when (t) {
//            is HttpException -> "网络不给力，请检查网络：" + t.message
//            is SocketTimeoutException -> "服务器响应超时，请检查网络"
//            is RuntimeException -> "运行时错误"
//            is UnknownHostException -> "无法解析主机，请检查网络"
//            is UnknownServiceException -> "未知的服务器错误，请检查网络"
//            else -> "网络不给力，请检查网络"
//        }
//        onFail(errorMsg)
//        disposable?.dispose()
//    }
//
//    override fun onComplete() {
//        disposable?.dispose()
//    }
//
//}
